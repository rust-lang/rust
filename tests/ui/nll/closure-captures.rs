// Some cases with closures that might be problems

// Should have one error per assignment

fn one_closure(x: i32) {
    ||
    x = 1; //~ ERROR
    move ||
    x = 1; //~ ERROR
}

fn two_closures(x: i32) {
    || {
        ||
        x = 1; //~ ERROR
    };
    move || {
        ||
        x = 1; //~ ERROR
    };
}

fn fn_ref<F: Fn()>(f: F) -> F { f }

fn two_closures_ref_mut(mut x: i32) {
    fn_ref(|| {
        || //~ ERROR
         x = 1;}
    );
    fn_ref(move || {
        ||  //~ ERROR
    x = 1;});
}

// This still gives two messages, but it requires two things to be fixed.
fn two_closures_ref(x: i32) {
    fn_ref(|| {
        || //~ ERROR
         x = 1;} //~ ERROR
    );
    fn_ref(move || {
        ||  //~ ERROR
    x = 1;}); //~ ERROR
}

fn two_closures_two_refs(x: &mut i32) {
    fn_ref(|| {
        || //~ ERROR
        *x = 1;});
    fn_ref(move || {
        || //~ ERROR
        *x = 1;});
}

fn main() {}
