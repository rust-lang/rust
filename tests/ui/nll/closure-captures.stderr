error[E0594]: cannot assign to `x`, as it is not declared as mutable
  --> $DIR/closure-captures.rs:7:5
   |
LL |     x = 1;
   |     ^^^^^ cannot assign
   |
help: consider changing this to be mutable
   |
LL | fn one_closure(mut x: i32) {
   |                +++

error[E0594]: cannot assign to `x`, as it is not declared as mutable
  --> $DIR/closure-captures.rs:9:5
   |
LL |     x = 1;
   |     ^^^^^ cannot assign
   |
help: consider changing this to be mutable
   |
LL | fn one_closure(mut x: i32) {
   |                +++

error[E0594]: cannot assign to `x`, as it is not declared as mutable
  --> $DIR/closure-captures.rs:15:9
   |
LL |         x = 1;
   |         ^^^^^ cannot assign
   |
help: consider changing this to be mutable
   |
LL | fn two_closures(mut x: i32) {
   |                 +++

error[E0594]: cannot assign to `x`, as it is not declared as mutable
  --> $DIR/closure-captures.rs:19:9
   |
LL |         x = 1;
   |         ^^^^^ cannot assign
   |
help: consider changing this to be mutable
   |
LL | fn two_closures(mut x: i32) {
   |                 +++

error[E0596]: cannot borrow `x` as mutable, as it is a captured variable in a `Fn` closure
  --> $DIR/closure-captures.rs:27:9
   |
LL | fn fn_ref<F: Fn()>(f: F) -> F { f }
   |                       - change this to accept `FnMut` instead of `Fn`
...
LL |     fn_ref(|| {
   |     ------ -- in this closure
   |     |
   |     expects `Fn` instead of `FnMut`
LL |         ||
   |         ^^ cannot borrow as mutable
LL |          x = 1;}
   |          - mutable borrow occurs due to use of `x` in closure

error[E0596]: cannot borrow `x` as mutable, as it is a captured variable in a `Fn` closure
  --> $DIR/closure-captures.rs:31:9
   |
LL | fn fn_ref<F: Fn()>(f: F) -> F { f }
   |                       - change this to accept `FnMut` instead of `Fn`
...
LL |     fn_ref(move || {
   |     ------ ------- in this closure
   |     |
   |     expects `Fn` instead of `FnMut`
LL |         ||
   |         ^^ cannot borrow as mutable
LL |     x = 1;});
   |     - mutable borrow occurs due to use of `x` in closure

error[E0594]: cannot assign to `x`, as it is not declared as mutable
  --> $DIR/closure-captures.rs:39:10
   |
LL |          x = 1;}
   |          ^^^^^ cannot assign
   |
help: consider changing this to be mutable
   |
LL | fn two_closures_ref(mut x: i32) {
   |                     +++

error[E0596]: cannot borrow `x` as mutable, as it is a captured variable in a `Fn` closure
  --> $DIR/closure-captures.rs:38:9
   |
LL | fn fn_ref<F: Fn()>(f: F) -> F { f }
   |                       - change this to accept `FnMut` instead of `Fn`
...
LL |     fn_ref(|| {
   |     ------ -- in this closure
   |     |
   |     expects `Fn` instead of `FnMut`
LL |         ||
   |         ^^ cannot borrow as mutable
LL |          x = 1;}
   |          - mutable borrow occurs due to use of `x` in closure

error[E0594]: cannot assign to `x`, as it is not declared as mutable
  --> $DIR/closure-captures.rs:43:5
   |
LL |     x = 1;});
   |     ^^^^^ cannot assign
   |
help: consider changing this to be mutable
   |
LL | fn two_closures_ref(mut x: i32) {
   |                     +++

error[E0596]: cannot borrow `x` as mutable, as it is a captured variable in a `Fn` closure
  --> $DIR/closure-captures.rs:42:9
   |
LL | fn fn_ref<F: Fn()>(f: F) -> F { f }
   |                       - change this to accept `FnMut` instead of `Fn`
...
LL |     fn_ref(move || {
   |     ------ ------- in this closure
   |     |
   |     expects `Fn` instead of `FnMut`
LL |         ||
   |         ^^ cannot borrow as mutable
LL |     x = 1;});
   |     - mutable borrow occurs due to use of `x` in closure

error[E0596]: cannot borrow `x` as mutable, as it is a captured variable in a `Fn` closure
  --> $DIR/closure-captures.rs:48:9
   |
LL | fn fn_ref<F: Fn()>(f: F) -> F { f }
   |                       - change this to accept `FnMut` instead of `Fn`
...
LL |     fn_ref(|| {
   |     ------ -- in this closure
   |     |
   |     expects `Fn` instead of `FnMut`
LL |         ||
   |         ^^ cannot borrow as mutable
LL |         *x = 1;});
   |         -- mutable borrow occurs due to use of `x` in closure

error[E0596]: cannot borrow `x` as mutable, as it is a captured variable in a `Fn` closure
  --> $DIR/closure-captures.rs:51:9
   |
LL | fn fn_ref<F: Fn()>(f: F) -> F { f }
   |                       - change this to accept `FnMut` instead of `Fn`
...
LL |     fn_ref(move || {
   |     ------ ------- in this closure
   |     |
   |     expects `Fn` instead of `FnMut`
LL |         ||
   |         ^^ cannot borrow as mutable
LL |         *x = 1;});
   |         -- mutable borrow occurs due to use of `x` in closure

error: aborting due to 12 previous errors

Some errors have detailed explanations: E0594, E0596.
For more information about an error, try `rustc --explain E0594`.
